# Data Clensing

## Prelude

In [1]:
import pandas as pd

DATA_FILE_CSV_PATH = "./flight_data.csv"

df_raw = pd.read_csv(f"{DATA_FILE_CSV_PATH}", sep=',', header=None, low_memory=False)

Take a glance at the data.

In [2]:
df_raw.head()

0      1             2            3           4               5   \
0  QUARTER  MONTH  DAY_OF_MONTH  DAY_OF_WEEK     FL_DATE  UNIQUE_CARRIER   
1        1      1             3            7  2016-01-03              F9   
2        1      1             3            7  2016-01-03              F9   
3        1      1             3            7  2016-01-03              F9   
4        1      1             3            7  2016-01-03              F9   

       6                  7                      8                      9   \
0  FL_NUM  ORIGIN_AIRPORT_ID  ORIGIN_AIRPORT_SEQ_ID  ORIGIN_CITY_MARKET_ID   
1     694              11292                1129202                  30325   
2     809              14027                1402702                  34027   
3     907              15356                1535602                  35356   
4     908              14492                1449202                  34492   

   ...        33       34        35             36             37         38  \
0  ...  AIR_TIME  FLIGHTS  DISTANCE  CARRIER_DELAY  WEATHER_DELAY  NAS_DELAY   
1  ...      87.0      1.0     692.0            NaN            NaN        NaN   
2  ...     224.0      1.0    1679.0           19.0            0.0        0.0   
3  ...      60.0      1.0     373.0            NaN            NaN        NaN   
4  ...      57.0      1.0     373.0            NaN            NaN        NaN   

               39                   40              41           42  
0  SECURITY_DELAY  LATE_AIRCRAFT_DELAY  FIRST_DEP_TIME  Unnamed: 42  
1             NaN                  NaN             NaN          NaN  
2             0.0                  0.0             NaN          NaN  
3             NaN                  NaN             NaN          NaN  
4             NaN                  NaN             NaN          NaN  

[5 rows x 43 columns]

Do you see *any* oddities? Column names are in fact in the first row! Let's name the columns as in the row and drop it.

In [3]:
df_raw.columns = df_raw.iloc[0]
df_raw.drop(df_raw.index[0], inplace=True)

See if any string type columns exist, which can be an nuisance for machine learning deployment.

In [4]:
df_raw.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1856061 entries, 1 to 1856061
Data columns (total 43 columns):
QUARTER                  object
MONTH                    object
DAY_OF_MONTH             object
DAY_OF_WEEK              object
FL_DATE                  object
UNIQUE_CARRIER           object
FL_NUM                   object
ORIGIN_AIRPORT_ID        object
ORIGIN_AIRPORT_SEQ_ID    object
ORIGIN_CITY_MARKET_ID    object
ORIGIN                   object
ORIGIN_CITY_NAME         object
ORIGIN_STATE_ABR         object
ORIGIN_STATE_NM          object
DEST_AIRPORT_ID          object
DEST_AIRPORT_SEQ_ID      object
DEST_CITY_MARKET_ID      object
DEST                     object
DEST_CITY_NAME           object
DEST_STATE_ABR           object
DEST_STATE_NM            object
CRS_DEP_TIME             object
DEP_TIME                 object
DEP_DELAY                object
DEP_DELAY_NEW            object
WHEELS_ON                object
TAXI_IN                  object
CRS_ARR_TIME           

Any odd zero (0) values anywhere?

In [5]:
df_raw.eq(0).sum()

0
QUARTER                  0
MONTH                    0
DAY_OF_MONTH             0
DAY_OF_WEEK              0
FL_DATE                  0
UNIQUE_CARRIER           0
FL_NUM                   0
ORIGIN_AIRPORT_ID        0
ORIGIN_AIRPORT_SEQ_ID    0
ORIGIN_CITY_MARKET_ID    0
ORIGIN                   0
ORIGIN_CITY_NAME         0
ORIGIN_STATE_ABR         0
ORIGIN_STATE_NM          0
DEST_AIRPORT_ID          0
DEST_AIRPORT_SEQ_ID      0
DEST_CITY_MARKET_ID      0
DEST                     0
DEST_CITY_NAME           0
DEST_STATE_ABR           0
DEST_STATE_NM            0
CRS_DEP_TIME             0
DEP_TIME                 0
DEP_DELAY                0
DEP_DELAY_NEW            0
WHEELS_ON                0
TAXI_IN                  0
CRS_ARR_TIME             0
ARR_TIME                 0
ARR_DELAY                0
ARR_DELAY_NEW            0
CRS_ELAPSED_TIME         0
ACTUAL_ELAPSED_TIME      0
AIR_TIME                 0
FLIGHTS                  0
DISTANCE                 0
CARRIER_DELAY            0

Great! None. Next, check on how pervasive Not A Number (NaN), converted from null value in the original data source upon importation into Pandas, are.

In [7]:
# Show the propertion of NaN in each and every column
# isnull() is an alias for isna(), so better to use isna()
(df_raw.isna().sum()/df_raw.shape[0]) * 100

0
QUARTER                    0.000000
MONTH                      0.000000
DAY_OF_MONTH               0.000000
DAY_OF_WEEK                0.000000
FL_DATE                    0.000000
UNIQUE_CARRIER             0.000000
FL_NUM                     0.000000
ORIGIN_AIRPORT_ID          0.000000
ORIGIN_AIRPORT_SEQ_ID      0.000000
ORIGIN_CITY_MARKET_ID      0.000000
ORIGIN                     0.000000
ORIGIN_CITY_NAME           0.000000
ORIGIN_STATE_ABR           0.000000
ORIGIN_STATE_NM            0.000000
DEST_AIRPORT_ID            0.000000
DEST_AIRPORT_SEQ_ID        0.000000
DEST_CITY_MARKET_ID        0.000000
DEST                       0.000000
DEST_CITY_NAME             0.000000
DEST_STATE_ABR             0.000000
DEST_STATE_NM              0.000000
CRS_DEP_TIME               0.000000
DEP_TIME                   1.429048
DEP_DELAY                  1.429048
DEP_DELAY_NEW              1.429048
WHEELS_ON                  1.512019
TAXI_IN                    1.512019
CRS_ARR_TIME              

Take a note of the followings:

1. Unusually high number of NaNs in the last seven (7) columns! Probably irrelevant and get rid of all of them!

In [8]:
df_raw.iloc[:,-7:]

0       CARRIER_DELAY WEATHER_DELAY NAS_DELAY SECURITY_DELAY  \
1                 NaN           NaN       NaN            NaN   
2                19.0           0.0       0.0            0.0   
3                 NaN           NaN       NaN            NaN   
4                 NaN           NaN       NaN            NaN   
5                 2.0           0.0      18.0            0.0   
...               ...           ...       ...            ...   
1856057           NaN           NaN       NaN            NaN   
1856058           NaN           NaN       NaN            NaN   
1856059           NaN           NaN       NaN            NaN   
1856060           NaN           NaN       NaN            NaN   
1856061           NaN           NaN       NaN            NaN   

0       LATE_AIRCRAFT_DELAY FIRST_DEP_TIME Unnamed: 42  
1                       NaN            NaN         NaN  
2                       0.0            NaN         NaN  
3                       NaN            NaN         NaN  
4                       NaN            NaN         NaN  
5                       0.0            NaN         NaN  
...                     ...            ...         ...  
1856057                 NaN            NaN         NaN  
1856058                 NaN            NaN         NaN  
1856059                 NaN            NaN         NaN  
1856060                 NaN            NaN         NaN  
1856061                 NaN            NaN         NaN  

[1856061 rows x 7 columns]

They don't look THAT relevant, don't they? We will drop them all.

In [9]:
df_raw.drop(df_raw.columns[-7:], axis=1, inplace=True)

2. DEP_TIME, DEP_DELAY and DEP_DELAY_NEW all have exactly the SAME 1.429048%, as well as WHEELS_ON, TAXI_IN and ARR_TIME (1.512019), and ARR_DELAY, ARR_DELAY_NEW, ACTUAL_ELAPSED_TIME and AIR_TIME (1.705655). Check if each and every row has NaN all across in a given set of columns.

In [23]:
print(df_raw[df_raw['DEP_TIME'].isna() & df_raw['DEP_DELAY'].isna() & df_raw['DEP_DELAY_NEW'].isna()].shape[0])
print(df_raw[df_raw['WHEELS_ON'].isna() & df_raw['TAXI_IN'].isna() & df_raw['ARR_TIME'].isna()].shape[0])
print(df_raw[df_raw['ARR_DELAY'].isna() & df_raw['ARR_DELAY_NEW'].isna() & df_raw['ACTUAL_ELAPSED_TIME'].isna() & df_raw['AIR_TIME'].isna()].shape[0])

26524
28064
31658


Let's count each column just to make sure.

In [19]:
print("1. 'DEP_TIME' = {0}, 'DEP_DELAY' = {1} & 'DEP_DELAY_NEW' = {2}".format(df_raw['DEP_TIME'].isna().sum(), df_raw['DEP_DELAY'].isna().sum(), df_raw['DEP_DELAY_NEW'].isna().sum()))
print("2. 'WHEELS_ON' = {0}, 'TAXI_IN' = {1} & 'ARR_TIME' = {2}".format(df_raw['WHEELS_ON'].isna().sum(), df_raw['TAXI_IN'].isna().sum(), df_raw['ARR_TIME'].isna().sum()))
print("3. 'ARR_DELAY' = {0}, 'ARR_DELAY_NEW' = {1}, 'ACTUAL_ELAPSED_TIME' = {2} & 'AIR_TIME' = {3}".format(df_raw['ARR_DELAY'].isna().sum(), df_raw['ARR_DELAY_NEW'].isna().sum(), df_raw['ACTUAL_ELAPSED_TIME'].isna().sum(), df_raw['AIR_TIME'].isna().sum()))

'DEP_TIME' = 26524, 'DEP_DELAY' = 26524 & 'DEP_DELAY_NEW' = 26524
'WHEELS_ON' = 28064, 'TAXI_IN' = 28064 & 'ARR_TIME' = 28064
'ARR_DELAY' = 31658, 'ARR_DELAY_NEW' = 31658, 'ACTUAL_ELAPSED_TIME' = 31658 & 'AIR_TIME' = 31658


Thus, from the above we can confidently eliminate info-redundant columns, i.e., leaving only one of those columns in each group.

The question is what column NAME to drop. For example, 'DEP_DELAY' and 'DEP_DELAY_NEW' are roughly the same meaning, whereas 'DEP_TIME' could be completely a disparate variable. Likewise, 'WHEELS_ON' and 'TAXI_IN' are identical and 

In [40]:
df_raw[df_raw.isnull().any(axis=1)]

0       QUARTER MONTH DAY_OF_MONTH DAY_OF_WEEK     FL_DATE UNIQUE_CARRIER  \
552           1     1            5           2  2016-01-05             F9   
553           1     1            5           2  2016-01-05             F9   
554           1     1            5           2  2016-01-05             F9   
555           1     1            5           2  2016-01-05             F9   
556           1     1            5           2  2016-01-05             F9   
...         ...   ...          ...         ...         ...            ...   
1854673       4    12            3           6  2016-12-03             NK   
1854681       4    12           11           7  2016-12-11             NK   
1854711       4    12           10           6  2016-12-10             NK   
1854717       4    12           16           5  2016-12-16             NK   
1854988       4    12           28           3  2016-12-28             DL   

0       FL_NUM ORIGIN_AIRPORT_ID ORIGIN_AIRPORT_SEQ_ID ORIGIN_CITY_MARKET_ID  \
552       1124             14771               1477102                 32457   
553       1125             12889               1288903                 32211   
554       1226             12889               1288903                 32211   
555       1227             13342               1334205                 33342   
556       1240             12889               1288903                 32211   
...        ...               ...                   ...                   ...   
1854673    121             13930               1393004                 30977   
1854681    121             13930               1393004                 30977   
1854711    124             14635               1463502                 31714   
1854717    124             14635               1463502                 31714   
1854988   2565             13487               1348702                 31650   

0        ... TAXI_IN CRS_ARR_TIME ARR_TIME ARR_DELAY ARR_DELAY_NEW  \
552      ...     NaN         2155      NaN       NaN           NaN   
553      ...     NaN         1925      NaN       NaN           NaN   
554      ...     NaN         1940      NaN       NaN           NaN   
555      ...     NaN         2225      NaN       NaN           NaN   
556      ...     NaN         1940      NaN       NaN           NaN   
...      ...     ...          ...      ...       ...           ...   
1854673  ...     NaN         1050      NaN       NaN           NaN   
1854681  ...     NaN         1050      NaN       NaN           NaN   
1854711  ...     NaN         1905      NaN       NaN           NaN   
1854717  ...     NaN         1907      NaN       NaN           NaN   
1854988  ...    11.0         2014   2230.0       NaN           NaN   

0       CRS_ELAPSED_TIME ACTUAL_ELAPSED_TIME AIR_TIME FLIGHTS DISTANCE  
552                100.0                 NaN      NaN     1.0    414.0  
553                105.0                 NaN      NaN     1.0    414.0  
554                215.0                 NaN      NaN     1.0   1524.0  
555                240.0                 NaN      NaN     1.0   1524.0  
556                185.0                 NaN      NaN     1.0   1222.0  
...                  ...                 ...      ...     ...      ...  
1854673            170.0                 NaN      NaN     1.0   1120.0  
1854681            170.0                 NaN      NaN     1.0   1120.0  
1854711            180.0                 NaN      NaN     1.0   1120.0  
1854717            183.0                 NaN      NaN     1.0   1120.0  
1854988            260.0                 NaN      NaN     1.0   1535.0  

[31658 rows x 36 columns]

Let's segregate into two (2) separate DataFrames: one to keep and the other to examine further on and after the 'DEP_TIME' column.

In [8]:
split_column_idx = df_raw.columns.get_loc('DEP_TIME')

df_keep = df_raw.iloc[:,:split_column_idx]
df_examine = df_raw.iloc[:,split_column_idx:]

In [9]:
df_examine.head()

0 DEP_TIME DEP_DELAY DEP_DELAY_NEW WHEELS_ON TAXI_IN CRS_ARR_TIME ARR_TIME  \
1   1524.0      -1.0           0.0    1807.0     8.0         1820   1815.0   
2    744.0      44.0          44.0     940.0     8.0          929    948.0   
3   1858.0      13.0          13.0    2006.0     7.0         2015   2013.0   
4   2054.0      -1.0           0.0    2208.0     7.0         2220   2215.0   
5   1252.0       2.0           2.0    1349.0    51.0         1420   1440.0   

0 ARR_DELAY ARR_DELAY_NEW CRS_ELAPSED_TIME ACTUAL_ELAPSED_TIME AIR_TIME  \
1      -5.0           0.0            115.0               111.0     87.0   
2      19.0          19.0            269.0               244.0    224.0   
3      -2.0           0.0             90.0                75.0     60.0   
4      -5.0           0.0             85.0                81.0     57.0   
5      20.0          20.0            150.0               168.0    107.0   

0 FLIGHTS DISTANCE  
1     1.0    692.0  
2     1.0   1679.0  
3     1.0    373.0  
4     1.0    373.0  
5     1.0    693.0

Need to conver 'object' to 'float' to run general arithmetic analysis.

In [10]:
pd.options.mode.chained_assignment = None  # default='warn'
# Otherwise the following warning appear
# /home/to/miniconda3/envs/us_flight_delay/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
# A value is trying to be set on a copy of a slice from a DataFrame.
# Try using .loc[row_indexer,col_indexer] = value instead
#
# See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
#   This is separate from the ipykernel package so we can avoid doing imports until

for col in df_examine:
    df_examine[col] = df_examine[col].astype(float)
    # Alternatively, df_examine[col] = pd.to_numeric(df_examine[col], errors='coerce')

df_examine.describe()

0          DEP_TIME     DEP_DELAY  DEP_DELAY_NEW     WHEELS_ON       TAXI_IN  \
count  1.829537e+06  1.829537e+06   1.829537e+06  1.827997e+06  1.827997e+06   
mean   1.334419e+03  9.812064e+00   1.267238e+01  1.471640e+03  7.562686e+00   
std    5.004642e+02  4.166410e+01   4.064121e+01  5.270121e+02  6.047238e+00   
min    1.000000e+00 -6.000000e+01   0.000000e+00  1.000000e+00  1.000000e+00   
25%    9.200000e+02 -5.000000e+00   0.000000e+00  1.053000e+03  4.000000e+00   
50%    1.329000e+03 -2.000000e+00   0.000000e+00  1.510000e+03  6.000000e+00   
75%    1.741000e+03  7.000000e+00   7.000000e+00  1.914000e+03  9.000000e+00   
max    2.400000e+03  2.040000e+03   2.040000e+03  2.400000e+03  2.500000e+02   

0      CRS_ARR_TIME      ARR_TIME     ARR_DELAY  ARR_DELAY_NEW  \
count  1.856061e+06  1.827997e+06  1.824403e+06   1.824403e+06   
mean   1.494374e+03  1.476238e+03  4.206204e+00   1.259892e+01   
std    5.133614e+02  5.314493e+02  4.389959e+01   4.031154e+01   
min    1.000000e+00  1.000000e+00 -1.070000e+02   0.000000e+00   
25%    1.109000e+03  1.057000e+03 -1.400000e+01   0.000000e+00   
50%    1.522000e+03  1.514000e+03 -6.000000e+00   0.000000e+00   
75%    1.920000e+03  1.919000e+03  7.000000e+00   7.000000e+00   
max    2.359000e+03  2.400000e+03  2.028000e+03   2.028000e+03   

0      CRS_ELAPSED_TIME  ACTUAL_ELAPSED_TIME      AIR_TIME    FLIGHTS  \
count      1.856058e+06         1.824403e+06  1.824403e+06  1856061.0   
mean       1.463799e+02         1.408941e+02  1.169484e+02        1.0   
std        7.673182e+01         7.587500e+01  7.374921e+01        0.0   
min        2.000000e+01         1.400000e+01  7.000000e+00        1.0   
25%        9.000000e+01         8.400000e+01  6.200000e+01        1.0   
50%        1.270000e+02         1.220000e+02  9.800000e+01        1.0   
75%        1.790000e+02         1.740000e+02  1.480000e+02        1.0   
max        7.050000e+02         7.780000e+02  7.230000e+02        1.0   

0          DISTANCE  
count  1.856061e+06  
mean   8.523126e+02  
std    6.198460e+02  
min    2.500000e+01  
25%    3.910000e+02  
50%    6.790000e+02  
75%    1.096000e+03  
max    4.983000e+03

Mmmm, numbers in scientific notation are much harder to grasp... Let's set float to 2 decimal points.

In [11]:
pd.options.display.float_format = '{:.2f}'.format

In [12]:
df_examine.describe()

0       DEP_TIME  DEP_DELAY  DEP_DELAY_NEW  WHEELS_ON    TAXI_IN  \
count 1829537.00 1829537.00     1829537.00 1827997.00 1827997.00   
mean     1334.42       9.81          12.67    1471.64       7.56   
std       500.46      41.66          40.64     527.01       6.05   
min         1.00     -60.00           0.00       1.00       1.00   
25%       920.00      -5.00           0.00    1053.00       4.00   
50%      1329.00      -2.00           0.00    1510.00       6.00   
75%      1741.00       7.00           7.00    1914.00       9.00   
max      2400.00    2040.00        2040.00    2400.00     250.00   

0      CRS_ARR_TIME   ARR_TIME  ARR_DELAY  ARR_DELAY_NEW  CRS_ELAPSED_TIME  \
count    1856061.00 1827997.00 1824403.00     1824403.00        1856058.00   
mean        1494.37    1476.24       4.21          12.60            146.38   
std          513.36     531.45      43.90          40.31             76.73   
min            1.00       1.00    -107.00           0.00             20.00   
25%         1109.00    1057.00     -14.00           0.00             90.00   
50%         1522.00    1514.00      -6.00           0.00            127.00   
75%         1920.00    1919.00       7.00           7.00            179.00   
max         2359.00    2400.00    2028.00        2028.00            705.00   

0      ACTUAL_ELAPSED_TIME   AIR_TIME    FLIGHTS   DISTANCE  
count           1824403.00 1824403.00 1856061.00 1856061.00  
mean                140.89     116.95       1.00     852.31  
std                  75.87      73.75       0.00     619.85  
min                  14.00       7.00       1.00      25.00  
25%                  84.00      62.00       1.00     391.00  
50%                 122.00      98.00       1.00     679.00  
75%                 174.00     148.00       1.00    1096.00  
max                 778.00     723.00       1.00    4983.00

What's obvious is that the FLIGHTS column is populated 1 in every row, so we eliminate that.

## Outlier Elimination

In [14]:
df_examine.drop(['FLIGHTS'], axis=1, inplace=True)

No data set is immaculately ready for number crunching. Let's compare two techniques to remove outliers.

1. Tukey Fences, and
2. Z-Score

In [27]:
import numpy as np

def outliers_iqr(data):
    q1, q3 = np.percentile(data, [25, 75])
    iqr = q3 - q1
    lower_bound = q1 - (iqr * 1.5)
    upper_bound = q3 + (iqr * 1.5)
    return np.where((data > upper_bound) | (data < lower_bound))

#t = outliers_iqr(df_examine['DEP_DELAY'])

# outliers = []

# for col in df_examine:
#     print(col)
#     print(outliers_iqr(df_examine[col]))
#     outliers.append([col, outliers_iqr(df_examine[col])[0]])

dict_outliers = {}

for col in df_examine:
    dict_outliers[col] = outliers_iqr(df_examine[col])[0]

dict_outliers

{'DEP_TIME': array([], dtype=int64),
 'DEP_DELAY': array([], dtype=int64),
 'DEP_DELAY_NEW': array([], dtype=int64),
 'WHEELS_ON': array([], dtype=int64),
 'TAXI_IN': array([], dtype=int64),
 'CRS_ARR_TIME': array([], dtype=int64),
 'ARR_TIME': array([], dtype=int64),
 'ARR_DELAY': array([], dtype=int64),
 'ARR_DELAY_NEW': array([], dtype=int64),
 'CRS_ELAPSED_TIME': array([], dtype=int64),
 'ACTUAL_ELAPSED_TIME': array([], dtype=int64),
 'AIR_TIME': array([], dtype=int64),
 'DISTANCE': array([     18,      43,      50, ..., 1855946, 1855991, 1856016])}

Mmm, the 'DISTANCE' column looks a bit fishy. The first output element of tupple prints row numbers in suspect. Check if they are atypical.

In [28]:
def display_all(df):
    with pd.option_context("display.max_rows", 1000, "display.max_columns", 1000): 
        display(df)

# Check on ORIGIN and DEST columns by displaying ALL the columns
# https://www.airmilescalculator.com/distance/las-to-mia/ tells you its distance is indeed 2,174 miles. No problem.
display_all(df_raw.iloc[dict_outliers['DISTANCE'], :])

0       QUARTER MONTH DAY_OF_MONTH DAY_OF_WEEK     FL_DATE UNIQUE_CARRIER  \
19            1     1            3           7  2016-01-03             F9   
44            1     1            3           7  2016-01-03             F9   
51            1     1            3           7  2016-01-03             F9   
52            1     1            3           7  2016-01-03             F9   
188           1     1            4           1  2016-01-04             F9   
...         ...   ...          ...         ...         ...            ...   
1855920       4    12           30           5  2016-12-30             DL   
1855924       4    12           30           5  2016-12-30             DL   
1855947       4    12           30           5  2016-12-30             DL   
1855992       4    12           30           5  2016-12-30             DL   
1856017       4    12           30           5  2016-12-30             DL   

0       FL_NUM ORIGIN_AIRPORT_ID ORIGIN_AIRPORT_SEQ_ID ORIGIN_CITY_MARKET_ID  \
19        1232             12889               1288903                 32211   
44        1233             13303               1330303                 32467   
51        1180             12892               1289203                 32575   
52        1181             13204               1320402                 31454   
188       1180             12892               1289203                 32575   
...        ...               ...                   ...                   ...   
1855920   1901             10397               1039705                 30397   
1855924   1905             12173               1217302                 32134   
1855947   1929             10397               1039705                 30397   
1855992   1973             13830               1383002                 33830   
1856017   1997             14747               1474703                 30559   

0       ORIGIN ORIGIN_CITY_NAME ORIGIN_STATE_ABR ORIGIN_STATE_NM  \
19         LAS    Las Vegas, NV               NV          Nevada   
44         MIA        Miami, FL               FL         Florida   
51         LAX  Los Angeles, CA               CA      California   
52         MCO      Orlando, FL               FL         Florida   
188        LAX  Los Angeles, CA               CA      California   
...        ...              ...              ...             ...   
1855920    ATL      Atlanta, GA               GA         Georgia   
1855924    HNL     Honolulu, HI               HI          Hawaii   
1855947    ATL      Atlanta, GA               GA         Georgia   
1855992    OGG      Kahului, HI               HI          Hawaii   
1856017    SEA      Seattle, WA               WA      Washington   

0       DEST_AIRPORT_ID DEST_AIRPORT_SEQ_ID DEST_CITY_MARKET_ID DEST  \
19                13303             1330303               32467  MIA   
44                12889             1288903               32211  LAS   
51                13204             1320402               31454  MCO   
52                12892             1289203               32575  LAX   
188               13204             1320402               31454  MCO   
...                 ...                 ...                 ...  ...   
1855920           14747             1474703               30559  SEA   
1855924           14747             1474703               30559  SEA   
1855947           14747             1474703               30559  SEA   
1855992           12892             1289204               32575  LAX   
1856017           12478             1247803               31703  JFK   

0         DEST_CITY_NAME DEST_STATE_ABR DEST_STATE_NM CRS_DEP_TIME DEP_TIME  \
19             Miami, FL             FL       Florida         2315   2311.0   
44         Las Vegas, NV             NV        Nevada          900    858.0   
51           Orlando, FL             FL       Florida         2050   2043.0   
52       Los Angeles, CA             CA    California         1655   1651.0   
188          Orlando, FL             

Z-Scores identify any data points by 3 standard deviations either side from mean, i.e., < -3 Stdev and < 3 Stdev.

In [29]:
def outliers_z_score(data):
    threshold = 3
    mean = np.mean(data)
    std = np.std(data)
    z_scores = [(y - mean) / std for y in data]
    return np.where(np.abs(z_scores) > threshold)

dict_outliers = {}

for col in df_examine:
    dict_outliers[col] = outliers_z_score(df_examine[col])[0]

dict_outliers

/home/to/miniconda3/envs/us_flight_delay/lib/python3.7/site-packages/ipykernel_launcher.py:6: RuntimeWarning: invalid value encountered in greater
  


{'DEP_TIME': array([], dtype=int64),
 'DEP_DELAY': array([    320,     324,     338, ..., 1855276, 1855707, 1855909]),
 'DEP_DELAY_NEW': array([    320,     324,     338, ..., 1855276, 1855707, 1855909]),
 'WHEELS_ON': array([], dtype=int64),
 'TAXI_IN': array([      4,      22,      30, ..., 1855855, 1855946, 1855994]),
 'CRS_ARR_TIME': array([], dtype=int64),
 'ARR_TIME': array([], dtype=int64),
 'ARR_DELAY': array([    320,     324,     338, ..., 1855276, 1855707, 1855909]),
 'ARR_DELAY_NEW': array([    320,     324,     338, ..., 1855276, 1855707, 1855909]),
 'CRS_ELAPSED_TIME': array([   3887,    3888,    3889, ..., 1855330, 1855633, 1855754]),
 'ACTUAL_ELAPSED_TIME': array([   3887,    3888,    3889, ..., 1855330, 1855633, 1855946]),
 'AIR_TIME': array([   3887,    3888,    3889, ..., 1855330, 1855633, 1855754]),
 'DISTANCE': array([   3887,    3888,    3889, ..., 1850596, 1855224, 1855633])}

In [30]:
for key, value in dict_outliers:
    

0       QUARTER MONTH DAY_OF_MONTH DAY_OF_WEEK     FL_DATE UNIQUE_CARRIER  \
321           1     1            4           1  2016-01-04             F9   
325           1     1            4           1  2016-01-04             F9   
339           1     1            5           2  2016-01-05             F9   
340           1     1            5           2  2016-01-05             F9   
344           1     1            5           2  2016-01-05             F9   
...         ...   ...          ...         ...         ...            ...   
1854820       4    12           28           3  2016-12-28             DL   
1854897       4    12           28           3  2016-12-28             DL   
1855277       4    12           29           4  2016-12-29             DL   
1855708       4    12           30           5  2016-12-30             DL   
1855910       4    12           30           5  2016-12-30             DL   

0       FL_NUM ORIGIN_AIRPORT_ID ORIGIN_AIRPORT_SEQ_ID ORIGIN_CITY_MARKET_ID  \
321        668             14771               1477102                 32457   
325       1367             14107               1410702                 30466   
339       1223             13204               1320402                 31454   
340       1224             12889               1288903                 32211   
344       1457             10397               1039705                 30397   
...        ...               ...                   ...                   ...   
1854820   2418             12892               1289204                 32575   
1854897   2477             11057               1105703                 31057   
1855277    302             12892               1289204                 32575   
1855708   1720             13485               1348502                 33485   
1855910   1893             10257               1025702                 30257   

0       ORIGIN   ORIGIN_CITY_NAME ORIGIN_STATE_ABR ORIGIN_STATE_NM  \
321        SFO  San Francisco, CA               CA      California   
325        PHX        Phoenix, AZ               AZ         Arizona   
339        MCO        Orlando, FL               FL         Florida   
340        LAS      Las Vegas, NV               NV          Nevada   
344        ATL        Atlanta, GA               GA         Georgia   
...        ...                ...              ...             ...   
1854820    LAX    Los Angeles, CA               CA      California   
1854897    CLT      Charlotte, NC               NC  North Carolina   
1855277    LAX    Los Angeles, CA               CA      California   
1855708    MSN        Madison, WI               WI       Wisconsin   
1855910    ALB         Albany, NY               NY        New York   

0       DEST_AIRPORT_ID DEST_AIRPORT_SEQ_ID DEST_CITY_MARKET_ID DEST  \
321               11292             1129202               30325  DEN   
325               14771             1477102               32457  SFO   
339               12889             1288903               32211  LAS   
340               13204             1320402               31454  MCO   
344               12889             1288903               32211  LAS   
...                 ...                 ...                 ...  ...   
1854820           14869             1486903               34614  SLC   
1854897           11433             1143302               31295  DTW   
1855277           10721             1072102               30721  BOS   
1855708           10397             1039705               30397  ATL   
1855910           13487             1348702               31650  MSP   

0            DEST_CITY_NAME DEST_STATE_ABR  DEST_STATE_NM CRS_DEP_TIME  \
321              Denver, CO             CO       Colorado         2000   
325       San Francisco, CA             CA     California         1807   
339           Las Vegas, NV             NV         Nevada         1855   
340             Orlando, FL             FL        Florida         2355   
344           Las Vegas, NV           